a10networks/a10-octavia
# [STACK-1321, 1322]: Added delete_member and delete_hm flow in delete_pool flow

## Description
- **Severity Level**: Critical
- Health monitor configuration was not getting removed from vThunder when we delete its associated Pool from openstack, even though, db entry for `health_monitor` gets removed.
- Delete a pool that has a member inside, will succeed in Octavia. But on Thunder, the server(member) is still there.

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-1321
https://a10networks.atlassian.net/browse/STACK-1322

## Technical Approach
- When pool is deleted, the child `members` and `hm` are deleted before.
- Invoke `delete_member_vthunder_subflow` flow inside `delete_pool_flow` flow in `a10_pool_flows.py`, if members are present.
- If the very last member of the project is deleted, the `delete_vrid_subflow` is also invoked. This is taken care by `mem_count` in 
`controller_worker` module.
- Invoke `delete_health_monitor_vthunder_subflow` flow inside `delete_pool` flow, if there exist an hm associated with that pool.
- Since the flow for `pool`, `member` and `hm` is same for `hardware_thunder` and `automated vthunder flows`, it will work in similar way.

## Config Changes
None

## Test Cases
1) **Normal Env**
- When pool is deleted via openstack cli command, the `health_monitor` associated with it are also deleted from vthunder config.
- When pool is deleted via openstack cli command, the `members` are also deleted from vthunder.
- Quotas in `quotas` table is correct before, during and after cli commands for `in_use_member`, `in_use_pool` and `in_use_health_monitor`.

2) **VRID config Env**
Config settings eg:
```
[a10_global]
vrid_floating_ip = 'dhcp'
```

- When pool is deleted via openstack cli command, the `health_monitor` associated with it are also deleted from vthunder config.
- When pool is deleted via openstack cli command, the `members` are also deleted from vthunder.
- VRID configuration corresponding to these members, is also deleted configuration from vthunder, in case the very last member in the `pool.project_id` is also deleted.
- Quotas in `quotas` table is correct before, during and after cli commands  for `in_use_member`, `in_use_pool` and `in_use_health_monitor`.

3) **VLAN config Env**
Config settings eg:
```
[a10_global]
network_type= 'vlan'

[hardware_thunder]
 devices=[
 {
 "project_id": "<>",
 "ip_address": "10.0.0.99",
 "device_name": "rack_vthunder",
 "username": "<>",
 "password": "<>",
 "interface_vlan_map": {
  "device_1": {
    "vcs_device_id": 1,
    "mgmt_ip_address": "10.0.0.74",
      "ethernet_interfaces": [{
        "interface_num": 2,
          "vlan_map": [{"vlan_id": 12, "ve_ip" : ".15"}]
   }],
   "trunk_interfaces": [{
    "interface_num": 1,
    "vlan_map": [
     {"vlan_id": 13, "use_dhcp": "True"},
     {"vlan_id": 14, "use_dhcp": "True"}
    ]
   }],
  },
<<similar device 2 config>>
```
- When pool is deleted via openstack cli command, the `health_monitor` associated with it are also deleted from vthunder config.
- When pool is deleted via openstack cli command, the `members` are also deleted from vthunder.
- VLAN interfaces corresponding to these members, are also deleted with ve interfaces configuration on vthunder.
- Quotas in `quotas` table is correct before, during and after cli commands  for `in_use_member`, `in_use_pool` and `in_use_health_monitor`.

## Manual Testing

In a complete SLB tree created environment, following scenarios occur in different environment:

1) **Normal Environment**
- `openstack loadbalancer pool delete pool1`
- Run `show running-config` in vthunder cli
- Check on vThunder, along with `service-group`, associated `health_monitor` is deleted
- On vthunder, all the associated `servers` are deleted.
- Check mysql table : `select * from quotas;` and check  `in_use_member`, `in_use_pool` and `in_use_health_monitor` column values.

2) **With VRID configured Environment**
- `openstack loadbalancer pool delete pool1`
- Run `show running-config` in vthunder cli
- Check on vThunder, along with `service-group`, associated `health_monitor` is deleted
- On vthunder, all the associated `servers` are deleted.
- `vrid` config is also deleted, if the very last `server` associated with that project is deleted.
- Check mysql table : `select * from quotas;` and check  `in_use_member`, `in_use_pool` and `in_use_health_monitor` column values.

3) **With VLAN configured environment**
- `openstack loadbalancer pool delete pool1`
- Run `show running-config` in vthunder cli
- Check on vThunder, `service-group` is deleted.
-  The associated `health_monitor` is also deleted.
- All the associated `server`'s are deleted.
- All Member related `vlan interface`  and `interface ve` are also deleted.
- Check mysql table : `select * from quotas;` and check  `in_use_member`, `in_use_pool` and `in_use_health_monitor` column values.

## Scripts for testing
```
openstack loadbalancer create --vip-subnet-id 91e9b009-4b7c-421b-b5ee-646fc86187db --name SLB1
sleep 6
openstack loadbalancer listener create  --protocol HTTP --protocol-port 80 SLB1 --name listener1
openstack loadbalancer pool create  --protocol HTTP --listener listener1 --lb-algorithm ROUND_ROBIN --name pool1
openstack loadbalancer member create --address 10.0.13.120 --subnet-id 2e3662b5-13bd-4f2c-a6dd-6006878dd791 --protocol-port 80 pool1 --name member1
sleep 6
openstack loadbalancer member create --address 10.0.13.124 --subnet-id 2e3662b5-13bd-4f2c-a6dd-6006878dd791 --protocol-port 80 pool1 --name member2
```
